Add scheduled sale fields

mattsherman opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR adds scheduled sale support to the new product editing experience.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes 43-gh-woocommerce/mothra-private

### How to test the changes in this Pull Request:

1. Install the [WC-Admin Test Helper](https://github.com/woocommerce/woocommerce-admin-test-helper) and enable the `new-product-management-experience`.
2. Go to `Products > Add New (MVP)`.
3. Go to the Pricing section.
4. Verify that acceptance criteria from 43-gh-woocommerce/mothra-private is met.

#### Known issues

- ~After saving a product with a scheduled sale, the product will be marked as dirty immediately (causing the "Update" button to be enabled). This was fixed by https://github.com/woocommerce/woocommerce/pull/35397~
- ~If you change the input of the scheduled to or from fields to the same date by typing, two onChange events get fired. This baffles me (none should be fired in this case). I will create an issue for follow-up at some later point (it's pretty edge-case).~ This was fixed in 8229b6f28d6856f5d81089cd240d7dc93d3014d8
- If the input of a DateTimePickerControl field is blank, the picker shows the current date as selected (this appears to be a Gutenberg regression) -- this should be followed up separately with Gutenberg
- If the site's timezone is not the same as the browser's timezone, the date selected in the picker is incorrect -- this should be addressed in a followup issue/PR

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

mattsherman committed:
Add scheduled sale fields to product form

Implement scheduled sale toggle.

Use GMT sale date fields

Update scheduled sale placeholders to be more clear

Make scheduled sale from and to date-only

Use site's date format

mdperez86 commented on packages/js/components/src/form/form.tsx:
```diff
@@ -302,6 +302,29 @@ function FormComponent< Values extends Record< string, any > >(
 		[ initialValues.current, values ]
 	);
 
+	function getDateTimePickerControlProps< Value = Values[ keyof Values ] >(
```
I think props adapter should be placed in a different place. The responsibility of matching props that are custom for the particular field should be managed by a HoC in the way of `mapStateToProps` or something similar. Instead of creating a new adapter like this anytime we are using a new field control that doesn't match the `getInputProps` signature we should rather create a WoC that adapts the new field control's props to understand the `getInputProps` signature. [This controller pattern](https://react-hook-form.com/get-started/#IntegratingwithUIlibraries) is another way to deal with that problem.

mdperez86 commented on packages/js/components/src/form/form.tsx:
```diff
@@ -302,6 +302,29 @@ function FormComponent< Values extends Record< string, any > >(
+		onBlur: () => void;
+		help?: string | null;
+	} {
+		const value = _get( values, name );
```
In case the prev comment is not gonna be considered for this PR then why we are not reusing the `getInputProps` inside this function and just adapt its responding signature? Logics like `help`, `className`, etc... are been duplicated.

mdperez86 commented on plugins/woocommerce-admin/client/products/sections/pricing-section.tsx:
```diff
@@ -162,6 +194,46 @@ export const PricingSection: React.FC = () => {
+										'Sale end date and time',
+										'woocommerce'
+									) }
+									{ ...getProductDateTimePickerControlProps( {
```
Well, here we have the adapter of an adapter. This can be solved also with my first comment.

mdperez86 commented on plugins/woocommerce-admin/client/products/sections/utils.ts:
```diff
@@ -108,3 +108,28 @@ export const getInputControlProps = ( {
 	};
 };
+
+export const getDateTimePickerControlProps = ( {
```
I think this function is not required. This is a case where reusability increases the complexity of things. If the only prop is gonna change is className then I just rather create a const for that, like `dateTimePickerClassName` which help us reducing a lot lines of code while reusability is been mantained.

mattsherman committed:
Fix margins for scheduled sale fields

mattsherman committed:
Set date_on_sale_from_gmt to beginning of day and date_on_sale_to_gmt to end of day

mattsherman committed:
Add scheduled sale tooltip

mattsherman committed:
Format dates using site options

mattsherman committed:
Simplify scheduled sale price toggle disabled check

mattsherman committed:
Pull scheduled sale fields validation out into separate function

mattsherman committed:
Disable eslint error for ts-ignore

mattsherman committed:
Make control styles more specific to avoid bleeding into areas (like the rick text editor toolbar)

mattsherman committed:
Fix SCSS lint issues (didn't have Stylelint set up properly in VSCode)

joshuatf commented on plugins/woocommerce-admin/client/products/layout/product-section-layout.scss:
```diff
@@ -10,14 +10,15 @@
-			&:not(:first-child):not(.components-radio-control) {
-				margin-top: $gap-large - $gap-smaller;
-				margin-bottom: 0;
+				> .components-base-control,
```
Nice 👍

joshuatf commented on plugins/woocommerce-admin/client/products/sections/pricing-section.tsx:
```diff
@@ -56,6 +75,51 @@ export const PricingSection: React.FC = () => {
+	const { dateFormat, timeFormat } = useSelect( ( select ) => {
+		const { getOption } = select( OPTIONS_STORE_NAME );
+		return {
+			dateFormat: ( getOption( 'date_format' ) as string ) || 'F j, Y',
```
Not something we need to implement in this PR, but just thinking out loud that it would be nice to have defaults like this stored somewhere.

I know we have some functions around intervals in the date package that has this for intervals, but feels too specific for this particular use case.

https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/date/src/index.ts#L795

mattsherman replied:
Agreed. Maybe it could belong in`date`.

joshuatf commented on plugins/woocommerce-admin/client/products/sections/pricing-section.tsx:
```diff
@@ -91,6 +155,12 @@ export const PricingSection: React.FC = () => {
 	const salePriceProps = getInputProps( 'sale_price', currencyInputProps );
 
+	const dateTimePickerProps = {
+		className: 'woocommerce-product__date-time-picker',
```
Just a general comment/question, I notice sometimes we've been using `woocommerce-product__` and `woocommerce-product-form__`.  

I don't have a strong opinion on which to use, but we might want to decide collectively on which to use going forward.  No need to hold up this PR though 😄

mattsherman replied:
💯 `woocommerce-product-form__` feels more specific, but I too don't feel strongly. I'll keep this as is for this PR and we can always have a followup PR to address consistency across the board.

joshuatf reviewed (commented):
Left a couple comments/questions but nothing that needs to block this PR.

I had some trouble testing.  Whenever I select a date, I receive the following error:

```
Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
```

I'm happy to dive into this further if you're unable to reproduce on your end.

octaedro reviewed (commented):
Great job @mattsherman! The code looks mainly good, I just left a couple of small comments, but I see a js error after picking the first date, and the page explodes after that.
```
Uncaught Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
    at checkForNestedUpdates (react-dom.js?ver=17.0.1:23833:17)
    at scheduleUpdateOnFiber (react-dom.js?ver=17.0.1:21865:5)
    at dispatchAction (react-dom.js?ver=17.0.1:16169:7)
    at index.js?ver=1667407934:2:603775
    at onBlur (index.js?ver=1667407934:2:604340)
    at onToggle (index.js?ver=1667407934:2:597503)
    at components.js?ver=c22dbc95bc435f9eeee5:29467:7
    at HTMLUnknownElement.callCallback (react-dom.js?ver=17.0.1:3942:16)
    at Object.invokeGuardedCallbackDev (react-dom.js?ver=17.0.1:3991:18)
    at invokeGuardedCallback (react-dom.js?ver=17.0.1:4053:33)
```

![Screen Capture on 2022-11-02 at 14-01-16](https://user-images.githubusercontent.com/1314156/199555124-ac3471f5-0765-4455-b054-383314a2d82a.gif)

octaedro commented on plugins/woocommerce-admin/client/products/sections/pricing-section.tsx:
```diff
@@ -157,6 +227,95 @@ export const PricingSection: React.FC = () => {
+													type="external"
+													onClick={ () =>
+														recordEvent(
+															'add_product_learn_more',
```
Does this name adhere to the convention of:
`<source>_<context>_<optional subcontext>_<action>_<optional qualifier>`?

Maybe we could also fix [this event name](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/products/sections/pricing-section.tsx#L111) as well.

mattsherman replied:
I used the existing `add_product_learn_more` event name. Let's address these in a followup PR.

octaedro commented on plugins/woocommerce-admin/client/products/sections/pricing-section.tsx:
```diff
@@ -56,6 +75,51 @@ export const PricingSection: React.FC = () => {
+		};
+	} );
+
+	const onSaleScheduleToggleChange = ( value: boolean ) => {
```
Can we record when the toggle button is turned on/off?

mattsherman replied:
Good idea! Done in 390b41a.

mattsherman commented:
Moving back to In Progress to address feedback

mattsherman committed:
Add Tracks event for when schedule sale toggle is clicked

mattsherman commented:
@octaedro @joshuatf Sorry about the `Uncaught Error: Maximum update depth exceeded.` error... I don't know what happened. I "always" test before asking for a review. But, obviously something snuck in that last-minute and I failed to test. I can reproduce and will push a fix this afternoon.

mattsherman committed:
Close DateTimePickerControl's dropdown when blurring from input.

mattsherman committed:
Remove unused code from stories file

mattsherman committed:
Log when onChange is called for Controlled stories

mattsherman committed:
Tweak how DateTimePickControl's dropdown gets the date

mattsherman commented:
@joshuatf @octaedro Ready for re-review!

mattsherman committed:
Handle when input changes occur but the end date is the same

mattsherman committed:
Fixed missed linter error

mattsherman committed:
Fix failing Form test

joshuatf reviewed (commented):
>I "always" test before asking for a review. But, obviously something snuck in that last-minute and I failed to test. I can reproduce and will push a fix this afternoon.

Not a problem at all 😄  Happens to the best of us.  I think there's sometimes a caching issue that occurs which prevents certain issues from appearing until you switch back and forth to the offending branch.  Or just goblins in our computers; one of those.

This tested really well for me on the product page, with dates and validation working as expected and persisted on refresh.  The non-controlled examples in Storybook seem to have broken in this branch.  Could you take a look at those?  I'm unable to select a value by clicking on a calendar date:
<img width="389" alt="Screen Shot 2022-11-04 at 2 51 36 PM" src="https://user-images.githubusercontent.com/10561050/200080957-76a0cd15-9f71-4431-9246-c4b3ef10d1ca.png">

mattsherman commented:
@joshuatf Thanks for reviewing and testing this! Hmm... I'll have to look at the code later for the non-controlled stories.

mattsherman committed:
Fix DateTimePickerControl usage when uncontrolled

mattsherman commented:
> The non-controlled examples in Storybook seem to have broken in this branch

@joshuatf I've fixed this in [91fb8e4](https://github.com/woocommerce/woocommerce/pull/34538/commits/91fb8e4f4dd7e61b40e03afc299ccd355e5550ec)

Sorry about that regression! And, thanks for catching it!

I believe everything is a-okay now. I rebased to address a file conflict.

mattsherman commented:
@joshuatf I just noticed the failed Form test. I'm investigating. I'll ping you again when I have it fixed.

mattsherman committed:
Fix handling when blur happens before debounced onChange

mattsherman commented:
@joshuatf Okay, all test pass again, the Storybook stories all appear to be working correctly, as does the product screen. So, Hopefully this is ready for a re-review now!

There are a few known issues I've listed in the PR description that I think can be handled in follow up PRs.

joshuatf reviewed (approved):
LGTM!  Testing well in both the new product experience and the storybook examples.  Thanks for all the changes here 💯

mattsherman commented:
Thanks for the reviews, @joshuatf ! 🙌
